File: sdks/python/src/agent_workflow/_canonical.py
"""Canonical JSON Form (DD-15): RFC 8785 JCS via ``rfc8785`` + the content rules.

The content rules (no nulls, absent-vs-empty, semantic array order) are emitter
obligations enforced upstream in the model/builder — by the time a wire dict reaches
this module it contains no ``None`` values, and canonicalization never adds or drops
anything (canonical rule 4: an explicitly-present empty section is preserved).
Hand-rolling JCS number formatting is the DD-15-rejected alternative; ``rfc8785``
implements the ECMAScript serialization the contract pins (``1.0`` → ``1``).
"""

from __future__ import annotations

from typing import Any

import rfc8785


def canonical_bytes(value: Any) -> bytes:
    """RFC 8785 canonical UTF-8 bytes of a parsed JSON value."""
    return rfc8785.dumps(value)

File: sdks/typescript/src/canonical.ts
/**
 * Canonical JSON Form (DD-15): RFC 8785 JCS via the `canonicalize` reference impl.
 *
 * The content rules (no nulls, absent-vs-empty, semantic array order) are emitter
 * obligations satisfied upstream: `undefined`-valued/omitted keys never reach the wire
 * (`canonicalize` drops them, matching the no-nulls rule), and a present-empty section
 * (`{}`) is preserved (canonical rule 4). JCS number formatting is native to JS —
 * `Number.prototype.toString` IS the ECMAScript algorithm RFC 8785 specifies — but we
 * use the reference library (DD-15's "don't reinvent JCS") rather than hand-rolling
 * key-sorting and string escaping.
 */

import canonicalize from "canonicalize";

/** RFC 8785 canonical JSON string of a JSON value. */
export function canonicalString(value: unknown): string {
  const out = canonicalize(value as object);
  if (out === undefined) {
    throw new Error("cannot canonicalize an undefined value");
  }
  return out;
}
